x/dex/keeper/core.go

```diff
@@ -182,25 +182,32 @@ func (k Keeper) WithdrawCore(goCtx context.Context, msg *types.MsgWithdrawl, tok
 	}
 	totalReserve0ToRemove := sdk.ZeroInt()
 	totalReserve1ToRemove := sdk.ZeroInt()
+	feeTiers := k.GetAllFeeTier(ctx)
 
 	for i, feeIndex := range msg.FeeIndexes {
 		sharesToRemove := msg.SharesToRemove[i]
 		tickIndex := msg.TickIndexes[i]
 
-		feeValue, found := k.GetFeeTier(ctx, feeIndex)
-		if !found {
-			return types.ErrValidFeeIndexNotFound
+		// check that feeIndex is a valid index of the fee tier
+		if feeIndex >= uint64(len(feeTiers)) {
+			return sdkerrors.Wrapf(types.ErrValidFeeIndexNotFound, "(%d) does not correspond to a valid fee", feeIndex)
 		}
-		fee := feeValue.Fee
+
+		fee := feeTiers[feeIndex].Fee
 		lowerTickIndex := tickIndex - fee
 		upperTickIndex := tickIndex + fee
+		sharesId := CreateSharesId(token0, token1, tickIndex, feeIndex)
+
+		if sharesBalance := k.bankKeeper.GetBalance(ctx, callerAddr, sharesId).Amount; sharesBalance.LT(sharesToRemove) {
```

Comment:
changed this to check against total shares instead. the check is only needed to avoid a division by zero error later in the withdrawal calculation. if the user has an insufficient balance the error will be raised when burning the shares.